controllers/effective_dnspolicies_reconciler.go

```diff
@@ -0,0 +1,291 @@
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+)
+
+func NewEffectiveDNSPoliciesReconciler(client *dynamic.DynamicClient, scheme *runtime.Scheme) *EffectiveDNSPoliciesReconciler {
+	return &EffectiveDNSPoliciesReconciler{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+type EffectiveDNSPoliciesReconciler struct {
+	client *dynamic.DynamicClient
+	scheme *runtime.Scheme
+}
+
+func (r *EffectiveDNSPoliciesReconciler) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.reconcile,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.DNSPolicyGroupKind},
+			{Kind: &DNSRecordGroupKind},
+		},
+	}
+}
+
+func (r *EffectiveDNSPoliciesReconciler) reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("EffectiveDNSPoliciesReconciler")
+
+	policyTypeFilterFunc := dnsPolicyTypeFilterFunc()
+	policyAcceptedFunc := dnsPolicyAcceptedStatusFunc(state)
+
+	policies := lo.FilterMap(topology.Policies().Items(), policyTypeFilterFunc)
+
+	policyErrors := map[string]error{}
+
+	logger.V(1).Info("updating dns policies", "policies", len(policies))
+
+	for _, policy := range policies {
+		pLogger := logger.WithValues("policy", policy.GetLocator())
+
+		if policy.GetDeletionTimestamp() != nil {
+			pLogger.V(1).Info("policy marked for deletion, skipping")
+			continue
+		}
+
+		if accepted, _ := policyAcceptedFunc(policy); !accepted {
+			pLogger.V(1).Info("policy not accepted, skipping")
+			continue
+		}
+
+		listeners := r.listenersForPolicy(ctx, topology, policy, policyTypeFilterFunc)
+
+		if logger.V(1).Enabled() {
+			listenerLocators := lo.Map(listeners, func(item *machinery.Listener, _ int) string {
+				return item.GetLocator()
+			})
+			pLogger.V(1).Info("reconciling policy for gateway listeners", "listeners", listenerLocators)
+		}
+
+		var gatewayHasAttachedRoutes = false
+		var gatewayHasAddresses = false
+
+		clusterID, err := utils.GetClusterUID(ctx, r.client)
+		if err != nil {
+			return fmt.Errorf("failed to generate cluster ID: %w", err)
+		}
+
+		for _, listener := range listeners {
+			lLogger := pLogger.WithValues("listener", listener.GetLocator())
+
+			gateway := listener.Gateway
+			if listener.Hostname == nil || *listener.Hostname == "" {
+				lLogger.Info("listener has no hostname assigned, skipping")
+				continue
+			}
+
+			if len(gateway.Status.Addresses) > 0 {
+				gatewayHasAddresses = true
+			}
+
+			hasAttachedRoute := false
+			for _, statusListener := range gateway.Status.Listeners {
+				if string(listener.Name) == string(statusListener.Name) {
+					hasAttachedRoute = statusListener.AttachedRoutes > 0
+				}
+			}
```

Comment:
This is existing code that has been around for a while https://github.com/Kuadrant/kuadrant-operator/blob/main/controllers/dnspolicy_dnsrecords.go#L83-L87. I would be surprised if you can have more than one status with the same section name, would kind of make it hard to know which one is related to what. It should possibly break when it finds it (assumes only one), or i guess do what you suggested and take the first positive one if there happened to be more.

Reply:
Yeah, this is a `status` per `Listener` defined in the `Gateway` so I guess it's possible to have multiple `Listeners` and only a `Route` attached to just one of the `Listener` using the route `sectionName` in the `parentRef` 🤔 

For example:
```yaml
      listeners:
        - attachedRoutes: 1
          conditions:
            - lastTransitionTime: "2024-10-23T08:54:09Z"
              message: No errors found
              observedGeneration: 1
              reason: Accepted
              status: "True"
              type: Accepted
            - lastTransitionTime: "2024-10-23T08:54:09Z"
              message: No errors found
              observedGeneration: 1
              reason: NoConflicts
              status: "False"
              type: Conflicted
            - lastTransitionTime: "2024-10-23T09:17:54Z"
              message: No errors found
              observedGeneration: 1
              reason: Programmed
              status: "True"
              type: Programmed
            - lastTransitionTime: "2024-10-23T09:17:54Z"
              message: No errors found
              observedGeneration: 1
              reason: ResolvedRefs
              status: "True"
              type: ResolvedRefs
          name: http
          supportedKinds:
            - group: gateway.networking.k8s.io
              kind: HTTPRoute
            - group: gateway.networking.k8s.io
              kind: GRPCRoute
        - attachedRoutes: 0
          conditions:
            - lastTransitionTime: "2024-10-23T08:54:09Z"
              message: No errors found
              observedGeneration: 1
              reason: Accepted
              status: "True"
              type: Accepted
            - lastTransitionTime: "2024-10-23T08:54:09Z"
              message: No errors found
              observedGeneration: 1
              reason: NoConflicts
              status: "False"
              type: Conflicted
            - lastTransitionTime: "2024-10-23T09:17:54Z"
              message: No errors found
              observedGeneration: 1
              reason: Programmed
              status: "True"
              type: Programmed
            - lastTransitionTime: "2024-10-23T09:17:54Z"
              message: No errors found
              observedGeneration: 1
              reason: ResolvedRefs
              status: "True"
              type: ResolvedRefs
          name: http2
          supportedKinds:
            - group: gateway.networking.k8s.io
              kind: HTTPRoute
            - group: gateway.networking.k8s.io
              kind: GRPCRoute

```
Since the `hasAttachedRoute` will always use the last `Listener` in the status block, this can be incorrect since this bool value also help to set `gatewayHasAttachedRoutes ` 🤔 

Since it's existing logic, it's probably a seperate issue (if it is) in a separate PR, but looks a bit odd to me at the moment anyway

Reply:
> Since the `hasAttachedRoute` will always use the last `Listener` in the status block, this can be incorrect since this bool value also help to set `gatewayHasAttachedRoutes ` 🤔

I'm still not really seeing the issue. Your status there has two listener status with different names, one with attached routes one without. The expected value of `gatewayHasAttachedRoutes` in this case will be `true` as we only set it to `true` when a listener [has attached routes](https://github.com/Kuadrant/kuadrant-operator/pull/937/files#diff-b6f99dd8a13a65bd856d8867f97c855b436c973a324732b8bc0e8a91f7663545R109-R111)  , default value is [false](https://github.com/Kuadrant/kuadrant-operator/pull/937/files#diff-b6f99dd8a13a65bd856d8867f97c855b436c973a324732b8bc0e8a91f7663545R87) 

`hasAttachedRoute` is set to true only when we find the listener with the desired name and [defaults to false](https://github.com/Kuadrant/kuadrant-operator/pull/937/files#diff-b6f99dd8a13a65bd856d8867f97c855b436c973a324732b8bc0e8a91f7663545R103).

I'd say create an issue if you think there is still an issue, there may well be one, but i can't see it currently.

I will however add a break.

Reply:
> hasAttachedRoute is set to true only when we find the listener with the desired name and [defaults to false](https://github.com/Kuadrant/kuadrant-operator/pull/937/files#diff-b6f99dd8a13a65bd856d8867f97c855b436c973a324732b8bc0e8a91f7663545R103).

Oh, I see now. I completely missed the fact that it does a match of listener status name with the desired name 🤦 In that case, it's not an issue. Ignore this 😅